coord/src/module.c

```diff
@@ -577,17 +769,139 @@ static void getReplyOffsets(const searchRequestCtx *ctx, searchReplyOffsets *off
     offsets->payload = 2;
     offsets->firstField = 3;
   }
-  if (ctx->withSortby || ctx->withSortingKeys) {
+
+  // Update the offsets for the special case after determining score, payload, field.
+  size_t specialCaseStartOffset = offsets->firstField;
+  size_t specialCasesMaxOffset = 0;
+  if(ctx->specialCases) {
+    size_t nSpecialCases = array_len(ctx->specialCases);
+    for(size_t i =0; i < nSpecialCases; i++) {
+      switch (ctx->specialCases[i]->specialCaseType)
+      {
+      case SPECIAL_CASE_KNN: {
+        ctx->specialCases[i]->knn.offset+=specialCaseStartOffset;
+        specialCasesMaxOffset = MAX(specialCasesMaxOffset, ctx->specialCases[i]->knn.offset);  
+        break;
+      }
+      case SPECIAL_CASE_SORTBY: {
+        ctx->specialCases[i]->sortby.offset += specialCaseStartOffset;
+        offsets->sortKey = ctx->specialCases[i]->sortby.offset;
+        specialCasesMaxOffset = MAX(specialCasesMaxOffset, ctx->specialCases[i]->sortby.offset);
+        break;
+      }
+      case SPECIAL_CASE_NONE:
+      default:
+        break;
+      }
+    }
+  }
+
+  if(specialCasesMaxOffset > 0) {
+    offsets->firstField=specialCasesMaxOffset+1;
+    offsets->step=offsets->firstField+1;
+  }
+  else if(ctx->withSortingKeys) {
     offsets->step++;
     offsets->sortKey = offsets->firstField++;
   }
+
   // nocontent - one less field, and the offset is -1 to avoid parsing it
   if (ctx->noContent) {
     offsets->step--;
     offsets->firstField = -1;
   }
 }
 
+
+/************************** Result processing callbacks **********************/
+
+static int cmp_scored_results(const void *p1, const void *p2, const void *udata) {
+  const scoredSearchResultWrapper* s1= p1;
+  const scoredSearchResultWrapper* s2 = p2;
+  double score1 = s1->score;
+  double score2 = s2->score;
+  if (score1 < score2) {
+    return -1;
+  } else if (score1 > score2) {
+    return 1;
+  }
+  return cmpStrings(s1->result->id, s1->result->idLen, s2->result->id, s2->result->idLen);
+}
+
+static void proccessKNNSearchReply(MRReply *arr, searchReducerCtx *rCtx, RedisModuleCtx *ctx) {
+  if (arr == NULL) {
+    return;
+  }
+  if (MRReply_Type(arr) == MR_REPLY_ERROR) {
+    rCtx->lastError = arr;
+    return;
+  }
+  if (MRReply_Type(arr) != MR_REPLY_ARRAY || MRReply_Length(arr) == 0) {
+    // Empty reply??
```

Comment:
What will happened here?

Reply:
empty reply? This is a copy from `processSearchReply` so it should be unified. let's discuss if needed

Reply:
Let verify what will happened here? Its not setting an error no?

Reply:
it should not. let's assume you have only 1 data point on 3 shards. 2 shards will return an empty response. Will add a test